dbms/src/Common/SimpleFixedThreadPool.cpp

```diff
@@ -0,0 +1,63 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Common/SimpleFixedThreadPool.h>
+#include <common/logger_useful.h>
+
+namespace DB
+{
+SimpleFixedThreadPool::SimpleFixedThreadPool(const std::string & name_, size_t m_size_)
+    : name(name_)
+    , m_size(m_size_)
+    , thread_mgr(newThreadManager())
+{
+    RUNTIME_ASSERT(m_size > 0);
+}
+
+void SimpleFixedThreadPool::schedule(ThreadManager::Job job)
+{
+    {
+        std::unique_lock lock(mutex);
+        has_free_thread.wait(lock, [this] { return active_jobs < m_size; });
+        ++active_jobs;
+    }
+    thread_mgr->schedule(false, name, [&, job]() {
+        job();
+        {
+            std::lock_guard lock(mutex);
+            --active_jobs;
+            has_free_thread.notify_one();
```

Comment:
I think we'd better calling `notify_one` out of the scope of lock

Reply:
https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one
```
The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock. However, some implementations (in particular many implementations of pthreads) recognize this situation and avoid this "hurry up and wait" scenario by transferring the waiting thread from the condition variable's queue directly to the queue of the mutex within the notify call, without waking it up.

Notifying while under the lock may nevertheless be necessary when precise scheduling of events is required, e.g. if the waiting thread would exit the program if the condition is satisfied, causing destruction of the notifying thread's condition variable. A spurious wakeup after mutex unlock but before notify would result in notify called on a destroyed object.
```
It doesn't seem to matter

Reply:
updated